src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -1109,221 +1096,274 @@ PeerImp::onMessage(std::shared_ptr<protocol::TMCluster> const& m)
+                    return badData("Invalid incomplete shard progress");
+                s.add32(progress);
             }
-            else
+
+            // Verify each incomplete shard is unique
+            if (!shardInfo.incomplete
+                     .emplace(
+                         shardIndex,
+                         NodeStore::ShardInfo::Incomplete(
+                             state, static_cast<float>(progress) / 100))
+                     .second)
             {
-                auto const earliestLedgerSeq{
-                    app_.getNodeStore().earliestLedgerSeq()};
-                earliestShard = NodeStore::seqToShardIndex(earliestLedgerSeq);
-                if (curLedgerSeq >= earliestLedgerSeq)
-                    latestShard = NodeStore::seqToShardIndex(curLedgerSeq);
+                return badData("Invalid duplicate incomplete shards");
             }
         }
-
-        if (boost::icl::first(shardIndexes) < earliestShard ||
-            (latestShard && boost::icl::last(shardIndexes) > latestShard))
-        {
-            return badData("Invalid shard indexes");
-        }
     }
 
-    // Get the IP of the node reporting the shard info
-    beast::IP::Endpoint endpoint;
-    if (m->has_endpoint())
+    // Verify final shards
+    if (m->has_final())
     {
-        if (m->endpoint() != "0")
+        auto const& final{m->final()};
+        if (final.empty())
+            return badData("Invalid final shards");
+
+        if (!from_string(shardInfo.final, final))
+            return badData("Invalid final shard indexes");
+
+        if (boost::icl::length(shardInfo.final) == 0 ||
+            boost::icl::first(shardInfo.final) < earliestIndex ||
+            (latestIndex && boost::icl::last(shardInfo.final) > latestIndex))
         {
-            auto result =
-                beast::IP::Endpoint::from_string_checked(m->endpoint());
-            if (!result)
-                return badData("Invalid incoming endpoint: " + m->endpoint());
-            endpoint = std::move(*result);
+            return badData("Invalid final shard indexes");
         }
-    }
-    else if (crawl())  // Check if peer will share IP publicly
-    {
-        endpoint = remote_address_;
+
+        s.addRaw(final.data(), final.size());
     }
 
-    // Get the Public key of the node reporting the shard info
-    PublicKey publicKey;
-    if (m->has_nodepubkey())
-        publicKey = PublicKey(makeSlice(m->nodepubkey()));
-    else
-        publicKey = publicKey_;
+    // Verify public key
+    auto slice{makeSlice(m->publickey())};
+    if (!publicKeyType(slice))
+        return badData("Invalid public key");
+
+    // Verify peer public key isn't this nodes's public key
+    PublicKey const publicKey(slice);
+    if (publicKey == app_.nodeIdentity().first)
+        return badData("Invalid public key");
 
+    // Verify signature
+    if (!verify(publicKey, s.slice(), makeSlice(m->signature()), false))
```

Comment:
Earlier I suggested you use `sign` instead of `signDigest` and noted that you'd have to change `verifyDigest` calls to `verify`. It seems you're ahead ;)

As is, this works by chance: you manually called **SHA512-Half** and passed that to `signDigest`. Here you don't hash the message, which should break. Luckily `verify` does that for you when using `secp256k1` keys.

So, long story short, you can leave this _as is_ and things will just work even after you make the change I described earlier, going from `signDigest` to `sign`.